servers/zms/src/main/java/com/yahoo/athenz/zms/store/impl/jdbc/JDBCConnection.java

```diff
@@ -1081,10 +1081,15 @@ public boolean insertDomainTags(String domainName, Map<String, StringList> tags)
             throw notFoundError(caller, ZMSConsts.OBJECT_DOMAIN, domainName);
         }
         int curTagCount = getDomainTagsCount(domainId);
-        int remainingTagsToInsert = domainTagsLimit - curTagCount;
+        int newTagCount = calculateTagCount(tags);
+        if (curTagCount + newTagCount > domainTagsLimit) {
+            throw requestError(caller, "domain tag quota exceeded - limit: "
+                + domainTagsLimit + ", current tags count: " + curTagCount + ", new tags count: " + newTagCount);
+        }
+
         boolean res = true;
         for (Map.Entry<String, StringList> e : tags.entrySet()) {
-            for (int i = 0; i < e.getValue().getList().size() && remainingTagsToInsert-- > 0; i++) {
+            for (int i = 0; i < e.getValue().getList().size(); i++) {
```

Comment:
since we no longer have the extra condition in the for loop, we can replace this with a simplified for loop:
for (String tagValue : e.getValue().getList())